src/core/Queue.cpp

```diff
@@ -193,87 +196,103 @@ bool Queue::isEmpty() const
-    {
-      cmd->waitList.pop_front();
-    }
-    else if (cmd->waitList.front()->state < 0)
+    Event *evt = command->waitList.front();
+    command->waitList.pop_front();
+
+    if (evt->state < 0)
     {
-      cmd->event->state = cmd->waitList.front()->state;
-      m_queue.pop();
-      return cmd;
+      command->event->state = evt->state;
+      return;
     }
-    else
+    else if (evt->state != CL_COMPLETE)
     {
-      return NULL;
+      evt->queue->execute(evt->command);
+      command->execBefore.push_front(evt->command);
     }
   }
 
-  cmd->event->startTime = now();
-  cmd->event->state = CL_RUNNING;
-
   // Dispatch command
-  switch (cmd->type)
+  command->event->startTime = now();
+  command->event->state = CL_RUNNING;
+
+  switch (command->type)
   {
-  case COPY:
-    executeCopyBuffer((CopyCommand*)cmd);
+  case Command::COPY:
+    executeCopyBuffer((CopyCommand*)command);
     break;
-  case COPY_RECT:
-    executeCopyBufferRect((CopyRectCommand*)cmd);
+  case Command::COPY_RECT:
+    executeCopyBufferRect((CopyRectCommand*)command);
     break;
-  case EMPTY:
+  case Command::EMPTY:
     break;
-  case FILL_BUFFER:
-    executeFillBuffer((FillBufferCommand*)cmd);
+  case Command::FILL_BUFFER:
+    executeFillBuffer((FillBufferCommand*)command);
     break;
-  case FILL_IMAGE:
-    executeFillImage((FillImageCommand*)cmd);
+  case Command::FILL_IMAGE:
+    executeFillImage((FillImageCommand*)command);
     break;
-  case READ:
-    executeReadBuffer((BufferCommand*)cmd);
+  case Command::READ:
+    executeReadBuffer((BufferCommand*)command);
     break;
-  case READ_RECT:
-    executeReadBufferRect((BufferRectCommand*)cmd);
+  case Command::READ_RECT:
+    executeReadBufferRect((BufferRectCommand*)command);
     break;
-  case KERNEL:
-    executeKernel((KernelCommand*)cmd);
+  case Command::KERNEL:
+    executeKernel((KernelCommand*)command);
     break;
-  case MAP:
-    executeMap((MapCommand*)cmd);
+  case Command::MAP:
+    executeMap((MapCommand*)command);
     break;
-  case NATIVE_KERNEL:
-    executeNativeKernel((NativeKernelCommand*)cmd);
+  case Command::NATIVE_KERNEL:
+    executeNativeKernel((NativeKernelCommand*)command);
     break;
-  case UNMAP:
-    executeUnmap((UnmapCommand*)cmd);
+  case Command::UNMAP:
+    executeUnmap((UnmapCommand*)command);
     break;
-  case WRITE:
-    executeWriteBuffer((BufferCommand*)cmd);
+  case Command::WRITE:
+    executeWriteBuffer((BufferCommand*)command);
     break;
-  case WRITE_RECT:
-    executeWriteBufferRect((BufferRectCommand*)cmd);
+  case Command::WRITE_RECT:
+    executeWriteBufferRect((BufferRectCommand*)command);
     break;
   default:
     assert(false && "Unhandled command type in queue.");
   }
 
-  cmd->event->endTime = now();
-  cmd->event->state = CL_COMPLETE;
+  command->event->endTime = now();
+  command->event->state = CL_COMPLETE;
+
+  // Remove command from its queue
+  m_queue.erase(it);
+}
+
+Command* Queue::update()
```

Comment:
For this kind of change of behaviour, I'd recommend renaming this method to something like `finish`, or removing it entirely and doing this directly inside `clFinish()` since it's so simple.

Reply:
Probably it's best to rename it to `finish()`. Placing this behaviour directly in `clFinish()` would require access to implementation details of the `Queue` class (namely access to `m_queue` in order to get the oldest command), breaking encapsulation.